File: app/src/test/java/com/dpis/module/root/RootAppProcessLauncherTest.java
package com.dpis.module.root;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;

import org.junit.Test;

public final class RootAppProcessLauncherTest {
    @Test
    public void shellQuoteEscapesSingleQuotes() {
        assertEquals(
                "'com.example/.MainActivity'",
                RootAppProcessLauncher.shellQuoteForTest("com.example/.MainActivity")
        );
        assertEquals("'a'\\''b'", RootAppProcessLauncher.shellQuoteForTest("a'b"));
    }

    @Test
    public void rootCommandsUseForceStopAndExplicitLauncherStart()
            throws IOException {
        String source = readSource("src/main/java/com/dpis/module/root/RootAppProcessLauncher.java");

        assertTrue(source.contains("am force-stop \" + packageName"));
        assertTrue(source.contains("am start --user current"));
        assertTrue(source.contains("-a android.intent.action.MAIN"));
        assertTrue(source.contains("-c android.intent.category.LAUNCHER"));
        assertTrue(source.contains("flattenToShortString()"));
        assertTrue(source.contains("isSafePackageName(packageName)"));
        assertTrue(source.contains("Runtime.getRuntime().exec(new String[] { \"su\", \"-c\", command })"));
    }

    private static String readSource(String relativePath) throws IOException {
        return new String(Files.readAllBytes(resolveSourcePath(relativePath)), StandardCharsets.UTF_8);
    }

    private static Path resolveSourcePath(String relativePath) throws IOException {
        Path current = Path.of("").toAbsolutePath();
        while (current != null) {
            Path fromModuleRoot = current.resolve(relativePath);
            if (Files.exists(fromModuleRoot)) {
                return fromModuleRoot;
            }
            Path fromRepositoryRoot = current.resolve("app").resolve(relativePath);
            if (Files.exists(fromRepositoryRoot)) {
                return fromRepositoryRoot;
            }
            current = current.getParent();
        }
        throw new IOException("source path not found: " + relativePath);
    }
}

File: app/src/test/java/com/dpis/module/root/RootAppProcessLauncherTest.kt
package com.dpis.module.root

import org.junit.Assert
import org.junit.Assert.assertEquals
import org.junit.Test
import java.io.IOException
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.nio.file.Path

class RootAppProcessLauncherTest {
    @Test
    fun shellQuoteEscapesSingleQuotes() {
        assertEquals(
            "'com.example/.MainActivity'",
            RootAppProcessLauncher.shellQuoteForTest("com.example/.MainActivity")
        )
        assertEquals("'a'\\''b'", RootAppProcessLauncher.shellQuoteForTest("a'b"))
    }

    @Test
    @Throws(IOException::class)
    fun rootCommandsUseForceStopAndExplicitLauncherStart() {
        val source: String =
            readSource("src/main/java/com/dpis/module/root/RootAppProcessLauncher.kt")

        Assert.assertTrue(source.contains("am force-stop \" + packageName"))
        Assert.assertTrue(source.contains("am start --user current"))
        Assert.assertTrue(source.contains("-a android.intent.action.MAIN"))
        Assert.assertTrue(source.contains("-c android.intent.category.LAUNCHER"))
        Assert.assertTrue(source.contains("flattenToShortString()"))
        Assert.assertTrue(source.contains("isSafePackageName(packageName)"))
        Assert.assertTrue(source.contains("SecureProcessLauncher.start(\"su\", \"-c\", command)"))
    }

    companion object {
        @Throws(IOException::class)
        private fun readSource(relativePath: String?): String {
            return String(
                Files.readAllBytes(resolveSourcePath(relativePath)),
                StandardCharsets.UTF_8
            )
        }

        @Throws(IOException::class)
        private fun resolveSourcePath(relativePath: String?): Path? {
            var current = Path.of("").toAbsolutePath()
            while (current != null) {
                val fromModuleRoot = current.resolve(relativePath)
                if (Files.exists(fromModuleRoot)) {
                    return fromModuleRoot
                }
                val fromRepositoryRoot = current.resolve("app").resolve(relativePath)
                if (Files.exists(fromRepositoryRoot)) {
                    return fromRepositoryRoot
                }
                current = current.parent
            }
            throw IOException("source path not found: " + relativePath)
        }
    }
}
